lib/nameMatcher.js

```diff
@@ -0,0 +1,307 @@
+        firstNamePromiseDev
+        ])
+        .then(function storeNamesLocally(names) {
+            // names = [lastNames, firstNames]
+            // firstNames = {bob: 'MP1:MP2', rachel: 'MP3:MP4'}
+            // assign names to local private variables
+            lastNames = parseMetaphones(names[0] || {});
+            firstNames = parseMetaphones(names[1] || {});
+            lastNamesDev = parseMetaphones(names[3] || {});
+            firstNamesDev = parseMetaphones(names[4] || {});
+
+            console.log('nameMatcher refreshing names complete');
+            return true;
+        }).catch(function catchRefreshError(err) {
+            console.error('Encountered error refreshing local names');
+            console.error(err);
+            throw err;
+        });
+};
+
+
+/**
+ * Initialization function. Sets redisClient, and creates whenReady promise that will resolve when the redis connection is ready and data has been populated from the redis DB;
+ * @return {object} nameMatcher
+ */
+var init = function () {
+    console.log('initializing nameMatcher');
+    nameMatcher.whenReady = new Promise(function promisifyWhenReady(resolve, reject) {
+        redisClient = redis.createClient();
+        redisClient.on('ready', function refreshWhenConnectionReady() {
+            console.log('nameMatcher redis connection established');
+            return refresh().then(function resolveWhenReadyPromise() {
+                console.log('nameMatcher initialization complete');
+                resolve(nameMatcher);
+            });
+        });
+    });
+    return nameMatcher;
+};
+
+
+// public methods
+
+/**
+ * Add a new name to the list of names (adds to in-memory list and to persistent storage)
+ * @param first {string} the first name
+ * @param last {string} the last name
+ * @return {Promise} a promise to be fulfilled with a boolean if both names were added (or are already existing)
+ */
+nameMatcher.addName = function (first, last, environment) {
+    if (environment === "prod") {
+        var firstMP = getMetaphone(first, firstNamesDev);
+        var lastMP = getMetaphone(last, lastNamesDev);
+        // Store the current values for the first and last name in case things go badly
+        var backupFirstName = firstNames[first];
+        var backupLastName = lastNames[last];
+        // Set names in local collections
+        firstNames[first] = firstMP;
+        lastNames[last] = lastMP;
+    } else if (environment === "dev") {
+        var firstMP = getMetaphone(first, firtNamesDev);
+        var lastMP = getMetaphone(last, lastNamesDev);
+        // Store the current values for the first and last name in case things go badly
+        var backupFirstName = firstNamesDev[first];
+        var backupLastName = lastNamesDev[last];
+        // Set names in local collections
+        firstNamesDev[first] = firstMP;
+        lastNamesDev[last] = lastMP;
+    }
+
+    var firstMPString = firstMP[0] + ':' + firstMP[1];
```

Comment:
maybe back up in `getMetaphones` you can provide an `e.g.` in the `@return` so I know what `firstMP[1]` should look like